jaxlib/pytree.cc

```diff
@@ -413,6 +415,150 @@ py::object PyTreeDef::Unflatten(py::iterable leaves) const {
   }
 }
 
+py::list PyTreeDef::FlattenUpTo(py::handle xs) const {
+  py::list leaves(num_leaves());
+  std::vector<py::object> agenda;
+  agenda.push_back(py::reinterpret_borrow<py::object>(xs));
+  auto it = traversal_.rbegin();
+  int leaf = num_leaves() - 1;
+  while (!agenda.empty()) {
+    if (it == traversal_.rend()) {
+      throw std::invalid_argument(
+          "Tree structures did not match; reached end of tree.");
+    }
+    const Node& node = *it;
+    py::object object = agenda.back();
+    agenda.pop_back();
+    ++it;
+
+    switch (node.kind) {
+      case Kind::kLeaf:
+        if (leaf < 0) {
+          throw std::logic_error("Leaf count mismatch.");
+        }
+        leaves[leaf] = py::reinterpret_borrow<py::object>(object);
+        --leaf;
+        break;
+
+      case Kind::kNone:
+        break;
+
+      case Kind::kTuple: {
+        if (!PyTuple_CheckExact(object.ptr())) {
+          throw std::invalid_argument(
+              absl::StrFormat("Expected tuple, got %s.", py::repr(object)));
+        }
+        py::tuple tuple = py::reinterpret_borrow<py::tuple>(object);
+        if (tuple.size() != node.arity) {
+          throw std::invalid_argument(
+              absl::StrFormat("Tuple arity mismatch: %d != %d; tuple: %s.",
+                              tuple.size(), node.arity, py::repr(object)));
+        }
+        for (py::handle entry : tuple) {
+          agenda.push_back(py::reinterpret_borrow<py::object>(entry));
+        }
+        break;
+      }
+
+      case Kind::kList: {
+        if (!PyList_CheckExact(object.ptr())) {
+          throw std::invalid_argument(
+              absl::StrFormat("Expected list, got %s.", py::repr(object)));
+        }
+        py::list list = py::reinterpret_borrow<py::list>(object);
+        if (list.size() != node.arity) {
+          throw std::invalid_argument(
+              absl::StrFormat("List arity mismatch: %d != %d; list: %s.",
+                              list.size(), node.arity, py::repr(object)));
+        }
+        for (py::handle entry : list) {
+          agenda.push_back(py::reinterpret_borrow<py::object>(entry));
+        }
+        break;
+      }
+
+      case Kind::kDict: {
+        if (!PyDict_CheckExact(object.ptr())) {
+          throw std::invalid_argument(
+              absl::StrFormat("Expected dict, got %s.", py::repr(object)));
+        }
+        py::dict dict = py::reinterpret_borrow<py::dict>(object);
+        py::list keys =
+            py::reinterpret_steal<py::list>(PyDict_Keys(dict.ptr()));
```

Comment:
It'd be nice to somehow combine the logic in this method with that in FlattenHelper, although I'm not sure how. Maybe some small helper methods for each Kind?

Reply:
Yeah I am not entirely sure how to do that. They are fundamentally doing slightly different things: one is traversing a Python term and deducing its structure, the other is iterating over a known structure and checking that a Python term matches. Leaving as is for now.